lib/src/gainmapmetadata.cpp

```diff
@@ -0,0 +1,305 @@
+    for (int c = 0; c < channelCount; ++c) {
+      streamReadU32(data, out_metadata->gainMapMinN[c], pos);
+      streamReadU32(data, out_metadata->gainMapMinD[c], pos);
+      streamReadU32(data, out_metadata->gainMapMaxN[c], pos);
+      streamReadU32(data, out_metadata->gainMapMaxD[c], pos);
+      streamReadU32(data, out_metadata->gainMapGammaN[c], pos);
+      streamReadU32(data, out_metadata->gainMapGammaD[c], pos);
+      streamReadU32(data, out_metadata->baseOffsetN[c], pos);
+      streamReadU32(data, out_metadata->baseOffsetD[c], pos);
+      streamReadU32(data, out_metadata->alternateOffsetN[c], pos);
+      streamReadU32(data, out_metadata->alternateOffsetD[c], pos);
+    }
+  }
+
+  // Fill the remaining values by copying those from the first channel.
+  for (int c = channelCount; c < 3; ++c) {
+    out_metadata->gainMapMinN[c] = out_metadata->gainMapMinN[0];
+    out_metadata->gainMapMinD[c] = out_metadata->gainMapMinD[0];
+    out_metadata->gainMapMaxN[c] = out_metadata->gainMapMaxN[0];
+    out_metadata->gainMapMaxD[c] = out_metadata->gainMapMaxD[0];
+    out_metadata->gainMapGammaN[c] = out_metadata->gainMapGammaN[0];
+    out_metadata->gainMapGammaD[c] = out_metadata->gainMapGammaD[0];
+    out_metadata->baseOffsetN[c] = out_metadata->baseOffsetN[0];
+    out_metadata->baseOffsetD[c] = out_metadata->baseOffsetD[0];
+    out_metadata->alternateOffsetN[c] = out_metadata->alternateOffsetN[0];
+    out_metadata->alternateOffsetD[c] = out_metadata->alternateOffsetD[0];
+  }
+
+  return JPEGR_NO_ERROR;
+}
+
+#define CHECK_NOT_ZERO(x)                \
+  do {                                   \
+    if (x == 0) {                        \
+      return ERROR_JPEGR_METADATA_ERROR; \
+    }                                    \
+  } while (0)
+
+status_t gain_map_metadata::gainmapMetadataFractionToFloat(const gain_map_metadata *from,
+                                                           ultrahdr_metadata_ptr to) {
+  if (from == nullptr || to == nullptr) {
+    return ERROR_JPEGR_BAD_PTR;
+  }
+
+  bool success = true;
+  CHECK_NOT_ZERO(from->baseHdrHeadroomD);
+  CHECK_NOT_ZERO(from->alternateHdrHeadroomD);
+  for (int i = 0; i < 3; ++i) {
+    CHECK_NOT_ZERO(from->gainMapMaxD[i]);
+    CHECK_NOT_ZERO(from->gainMapGammaD[i]);
+    CHECK_NOT_ZERO(from->gainMapMinD[i]);
+    CHECK_NOT_ZERO(from->baseOffsetD[i]);
+    CHECK_NOT_ZERO(from->alternateOffsetD[i]);
+  }
+  to->version = kGainMapVersion;
+  to->maxContentBoost = from->gainMapMaxN[0] / from->gainMapMaxD[0];
+  to->minContentBoost = from->gainMapMinN[0] / from->gainMapMinD[0];
+  to->gamma = from->gainMapGammaN[0] / from->gainMapGammaD[0];
+
+  // BaseRenditionIsHDR is false
+  to->offsetSdr = from->baseOffsetN[0] / from->baseOffsetD[0];
+  to->offsetHdr = from->alternateOffsetN[0] / from->alternateOffsetD[0];
+  to->hdrCapacityMax = from->alternateHdrHeadroomN / from->alternateHdrHeadroomD;
+  to->hdrCapacityMin = from->baseHdrHeadroomN / from->baseHdrHeadroomD;
```

Comment:
These are integer divisions (both numerator and denominator are uint32_t) so they will not produce the correct float value. You need to cast at least one of the two operands into a float.